Source/SecureChat/KeysRotation/KeysRotator.swift

```diff
@@ -0,0 +1,278 @@
+                    let longTermPublicKey = self.crypto.exportPublicKey(longTermKeyPair.publicKey)
+                    let longTermKeyId = try self.keyUtils.computePublicKeyId(publicKey: longTermPublicKey)
+                    _ = try self.longTermKeysStorage.storeKey(longTermPrivateKey,
+                                                              withId: longTermKeyId)
+                    let longTermKeySignature = try self.crypto.generateSignature(of: longTermPublicKey,
+                                                                                 using: self.identityPrivateKey)
+                    longTermSignedPublicKey = SignedPublicKey(publicKey: longTermPublicKey,
+                                                              signature: longTermKeySignature)
+                }
+                else {
+                    longTermSignedPublicKey = nil
+                }
+
+                let numOfrelevantOneTimeKeys = oneTimeKeysIds.count - validateResponse.usedOneTimeKeysIds.count
+                let numbOfOneTimeKeysToGen = UInt(max(self.desiredNumberOfOneTimeKeys - numOfrelevantOneTimeKeys, 0))
+
+                Log.debug("Generating \(numbOfOneTimeKeysToGen) one-time keys")
+                let oneTimePublicKeys: [Data]
+                if numbOfOneTimeKeysToGen > 0 {
+                    let keyPairs = try self.crypto.generateMultipleKeyPairs(numberOfKeyPairs: numbOfOneTimeKeysToGen)
+
+                    var publicKeys = [Data]()
+                    publicKeys.reserveCapacity(Int(numbOfOneTimeKeysToGen))
+                    for keyPair in keyPairs {
+                        let oneTimePrivateKey = self.crypto.exportPrivateKey(keyPair.privateKey)
+                        let oneTimePublicKey = self.crypto.exportPublicKey(keyPair.publicKey)
+                        let keyId = try self.keyUtils.computePublicKeyId(publicKey: oneTimePublicKey)
+                        _ = try self.oneTimeKeysStorage.storeKey(oneTimePrivateKey, withId: keyId)
+
+                        publicKeys.append(oneTimePublicKey)
+                    }
+
+                    oneTimePublicKeys = publicKeys
+                }
+                else {
+                    oneTimePublicKeys = []
+                }
+
+                Log.debug("Uploading keys")
+                try self.client.uploadPublicKeys(identityCardId: self.identityCardId,
+                                                 longTermPublicKey: longTermSignedPublicKey,
+                                                 oneTimePublicKeys: oneTimePublicKeys,
+                                                 token: token.stringRepresentation())
+
+                completionWrapper(Void(), nil)
+            }
+            catch {
+                completionWrapper(nil, error)
+            }
+        }
+    }
+}
+
+// MARK: - KeysRotatorProtocol
+extension KeysRotator: KeysRotatorProtocol { }
```

Comment:
Main declaration of class has required by protocol method. It should be moved to extension or extension should be deleted.

Reply:
I don't get your point. Could you explain?

Reply:
What's the point of making extension, which inherits protocol, but implements required methods in other place?

Reply:
OK, doesn't see much difference. Fixed